### PR TITLE
Improve env layer handling of models.

### PR DESCRIPTION
**Pull request by frankban**
Implement a method for listing detailed info on
existing models.
Add the ability to destroy the current model.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4878/
Test PASSed.

File: jujugui/static/gui/src/app/store/env/go.js [hunk 1]
```diff
@@ -729,39 +735,129 @@ YUI.add('juju-env-go', function(Y) {
           return;
         }
         var results = data.Response.results;
-        if (results.length !== 1) {
+        if (results.length !== tags.length) {
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: Could you add a comment here as to why this check is being made please.

File: jujugui/static/gui/src/app/store/env/go.js [hunk 2]
```diff
@@ -2921,19 +3017,55 @@ YUI.add('juju-env-go', function(Y) {
       callback(transformedData);
     },
 
+    /**
+      Destroy the current connected model.
+
+      For obvious reasons callers should switch the WebSocket connection to
```
Comment (hatched) [inline jujugui/static/gui/src/app/store/env/go.js]: s/For obvious reasons callers/Callers/

It's not obvious or intuitive :)

File: jujugui/static/gui/src/test/test_env_go.js [hunk 3]
```diff
@@ -710,9 +745,79 @@ with this program.  If not, see <http://www.gnu.org/licenses/>.
...
+
+    it('retrieves model info for multiple models', function(done) {
+      // Perform the request.
+      var tag1 = 'model-5bea955d-7a43-47d3-89dd-tag1';
+      var tag2 = 'model-5bea955d-7a43-47d3-89dd-tag2';
+      env.modelInfo([tag1, tag2], function(data) {
+        assert.strictEqual(data.err, undefined);
+        assert.strictEqual(data.models.length, 2);
+        var result1 = data.models[0];
+        assert.strictEqual(result1.tag, tag1);
```
Comment (hatched) [inline jujugui/static/gui/src/test/test_env_go.js]: Just for the future, you could have used 'deepEqual' here, would have saved some typing :) 
```JavaScript
assert.deepEqual(data.models, [{...}, {...}])
```

**Comment by hatched** [comment]
:+1: With some trivial comments
Looks good thanks!

File: jujugui/static/gui/src/app/store/env/go.js (hunk 1 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/go.js]: Done.

File: jujugui/static/gui/src/app/store/env/go.js (hunk 2 above)
Comment (frankban) [inline jujugui/static/gui/src/app/store/env/go.js]: Done.

File: jujugui/static/gui/src/test/test_env_go.js (hunk 3 above)
Comment (frankban) [inline jujugui/static/gui/src/test/test_env_go.js]: Yes, I reused the same pattern used elsewhere and didn't change it with the consideration that deepEquals, when it fails, produces a meaningless output.

**Comment by jujugui** [comment]
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4879/
Test FAILed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/juju-gui/4885/
Test PASSed.

File: jujugui/static/gui/src/app/store/env/go.js [hunk 4]
```diff
@@ -729,39 +735,130 @@ YUI.add('juju-env-go', function(Y) {
...
+        - isAlive: whether the model is alive or dying/dead;
+        - isAdmin: whether the model is an admin model;
+        - lastConnection: the date of the last connection as a string, e.g.:
+          '2015-09-24T10:08:50Z' or null if the model was never connected to;
+        - err: a message describing a specific model error, or undefined.
+      @return {undefined} Sends a message to the server only.
+    */
+    listModelsWithInfo: function(callback) {
+      // Ensure we always have a callback.
+      if (!callback) {
```
Comment (kadams54) [inline jujugui/static/gui/src/app/store/env/go.js]: For my own curiosity… I don't recall seeing this pattern in other functions that take callbacks. Is it something that should be more widespread?

**Comment by kadams54** [comment]
👍 from me as well!

**Comment by hatched** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/juju-gui-merge

**Pull request merged**